src/msw/srchctrl.cpp

```diff
@@ -0,0 +1,448 @@
+                
+                // In LTR layout, we can skip rightRect (cancel button) calculation.
+                if ( !(GetLayoutDirection() == wxLayout_LeftToRight && !IsCancelButtonVisible()) )
+                {
+                    // Calculate rightRect
+                    rightRect = rcWindow;
+                    rightRect.left = rightRect.right - bitmapWidth;
+
+                    ::OffsetRect(&rightRect, rcClient.right + bitmapWidth - rcWindow.right, 0);
+                    rightRect.top += rcClient.top - rcWindow.top;
+                    rightRect.bottom -= rcWindow.bottom - rcClient.bottom;
+
+                    ::OffsetRect(&rightRect, -rcWindow.left, -rcWindow.top);
+                }
+
+                // copy results
+                wxCopyRECTToRect(*searchButtonRect, m_searchButtonRect);
+                wxCopyRECTToRect(*cancelButtonRect, m_cancelButtonRect);
+
+                DrawButtons(bitmapWidth);
+            }
+            break;
+
+        case WM_NCHITTEST:
+            {
+                POINT pt;
+                pt.x = GET_X_LPARAM(lParam);
+                pt.y = GET_Y_LPARAM(lParam);
+                ::MapWindowPoints(NULL, GetHwnd(), &pt, 1);
+
+                const wxPoint point = wxPoint(pt.x + bitmapWidth, pt.y);
+
+                if ( IsCancelButtonVisible() && m_cancelButtonRect.Contains(point) )
+                {
+                    m_mouseInCancelButton = true;
+                    *rc = HTCAPTION;
+                    processed = true;
+
+                    TRACKMOUSEEVENT tme = { sizeof(tme), TME_NONCLIENT | TME_LEAVE, GetHwnd() };
+                    TrackMouseEvent(&tme);
+                }
+                else if ( m_searchButtonRect.Contains(point) )
+                {
+                    *rc = HasMenu() ? HTCAPTION : HTCLIENT;
+                    processed = true;
+                }
+            }
+            break;
+
+        case WM_NCLBUTTONDOWN:
+            {
+                ClickButton();
+                processed = true;
+            }
+            break;
+
+        case WM_NCMOUSELEAVE:
+            {
+                m_mouseInCancelButton = false;
+            }
+            wxFALLTHROUGH;
+
+        case WM_NCMOUSEMOVE:
+            {
+                HighlightButton();
+            }
+            break;
+
+            // Reduce noticeable flicker by processing WM_MOUSEMOVE here
+            // by doing nothing if it just a mouse move only.
+        case WM_MOUSEMOVE:
+            {
+                // ...text can be selected though.
+                if ( wParam & MK_LBUTTON )
+                    break;
+            }
+            wxFALLTHROUGH;
+
+        case WM_NCLBUTTONDBLCLK:
+            {
+                *rc = 0;
+                processed = true;
+            }
+            break;
+    }
+
+    return processed || wxTextCtrl::MSWHandleMessage(rc, nMsg, wParam, lParam);
+}
+
+void wxSearchCtrl::DrawButtons(int width)
+{
+    const double xWidth = 80.;
```

Comment:
Would be nice to have symbolic names for all the constants in this function too.

I also think that at least this one should use `FromDIP()`.

Reply:
> I also think that at least this one should use FromDIP().

No, it looks good on my high DPI screen. `gdc.SetUserScale(scale, scale);` takes care of this.

Reply:
Ah, I see, thanks. It would still be nice to have some names for them and maybe a comment explaining why we don't need `FromDPI()` here, just in case somebody overeagerly adds it later.